### ISSUE TITLE
Proposal: remove referee.count

### ISSUE THREAD

**Issue by mroderick** [sinonjs/referee#47]
`referee.count` is an artefact from `buster`, which was made so that `buster` could report on how many assertions were run, and could fail a test that had no assertions.

I don't know of any other test runner that has this feature or uses `referee.count` for anything.

Since `master` is already primed for another MAJOR release by merging #46, I think now would be a good time to remove `referee.count`.

Opinions?

**Comment by mantoni** [comment]
Actually, I never thought about this feature, but now that I'm reading your proposal, I'm thinking about using it 😆 

I can imagine to add something like this to each test:

```js
afterEach(() => {
  referee.refute.equals(referee.count, 0);
});
```

However, I'd rather add something more expressive to the API instead, e.g. `referee.verify()` or `assert.assertions()`.

I think the Tap test runner uses a similar feature. I'm not sure if it's possible to integrate Tab with Referee though.

**Comment by mroderick** [comment]
The problem with relying on assertion count is when you also use tools like `jsverify`, that does create Error instances (which your test runner expects). It doesn't have an assertion in `referee` which can be counted.

I remember adding token assertions to tests in `buster`, just so the test runner wouldn't complain.

**Comment by mantoni** [comment]
It would be something you could call optionally in `afterEach`, not baked into your test runner.

**Comment by mroderick** [comment]
> Maybe we can remove count officially, but somehow keep track of assert calls internally? I would like to try adding a method to verify asserts where made and to reset the counter. Makes sense?

I thought about that if we would keep it, we would make a few changes:

* make the actual value private, and expose it via a getter (prevent accidental and intentional writes to it)
* add a `resetCount` method
* add a `verifyCount` method

We should either have a great feature, or remove it. If we remove it, the library is stateless, which makes most things easier. If we keep it, we should mature it, so it's not possible to write to the count.

**Comment by mantoni** [comment]
Yes, that sounds like a good idea. There is something else I'm thinking about: If we're adding promise based tests with #45, you described a problem were it might be easy to have a passing test because the failure would happen asynchronously. We could introduce `assert.verify()` to check that `count` is not zero, reset the `count` and also verify that all promises have been resolved. I think it would be safe to mix both assertions because `count` is implicitly increased anyway when `assert.resolved` is used.

**Comment by mroderick** [comment]
Let's see if I've understood your comment:

* Create a verify method that verifies that `count` is not zero.
* If #45 lands, then expand `verify` to ensure all promises from `resolves` and `rejects` are resolved or rejected

I am not sure if this is sufficient to ensure that there won't be passing tests that shouldn't be.
If authors forget to use `.verify()` with `resolves`/`rejects` assertions, then they still won't know that their tests are no good. 

It's certainly an improvement, but is it enough?

We could detect when running in node, and hook into `process.on('exit')` and detect if there are unverified promises, and if not, then set a non-zero exit code. But, is that really the job of an assertion library?

**Comment by mroderick** [comment]
Thinking about it some more, maybe it's fine: https://mochajs.org/#working-with-promises

**Comment by mantoni** [comment]
You understood me correctly. That describes the plan.

I think adding a method is enough for an assertion library. I would think of this like calling `sinon.restore()`. It's optional, so yes, you can forget it, but at least we don't add magical behaviour.

As a side node: Hooking into `process.on('exit')` isn't guaranteed to work either because newer versions of Mocha don't call `process.exit()` anymore. They simply set `process.exitCode = 1` in case of a failure and let node decide whether it's done.

**Comment by mantoni** [comment]
Sure. The original request in this issue was addressed by #48. We can continue with a followup issue for #45 to enhance `verify`. Let's close this one 👍
